### ISSUE TITLE
Not running as DaemonSet?

### ISSUE THREAD

**Issue by rickypai** [jtblin/kube2iam#8]
One of the concerns of running kube2iam as a container on each host through DaemonSet is that it requires all minion nodes to have the `sts:AssumeRole` permission. This means any escalation to the host network or container interface, intended or otherwise, has the potential of gaining all application roles.

Since `kube2iam` has dependencies on only the Kubernetes API and the EC2 metadata service, both reachable by network, it seems like, with some modifications, we can run kube2iam on dedicated nodes. This way, we restrict the number of EC2 instances having `sts:AssumeRole` permission, reducing the exposure of the potentially sensitive permission.

**Comment by jtblin** [comment]
Hi @rickypai, makes sense. I think you should not have any problem to run it on a separate server, the proxy already supports `--api-server` and `--api-token` which you would need to authenticate with the kubernetes server. You would need to update the iptables rule to point to your external server etc.
If any further changes are needed, I am more than happy to accept pull requests or to do required changes. Also if you set it up this way, it would be great if you could let me know of the steps etc. and we can update the README.md to show the configuration needed to accomplish that.

**Comment by rickypai** [comment]
I've been experimenting with this setup by first running kube2iam as Daemonset with hostNetwork, but in my iptables rule, instead of using local IP for destination, a Kubernetes service IP.

I want to do this first by validating that we could get kube2iam working through exposing as a Kubernetes service, then potentially run kube2iam as static pods only on the master nodes. This way, we'd limit the use of `sts:AssumeRole` on just the master nodes.

Problems I ran into:
#### iptables routing doesn't seem to work to a service IP
the routing rule doesn't seem to work. I suspect this is because I need to bridge between docker0 and flannel.1? but it gives me this error: `iptables v1.4.21: Can't use -o with PREROUTING`. I suspect I'm doing things wrong or need a more complicated routing rule. This is the command I'm using:

```
sudo iptables \
  --table nat \
  --append PREROUTING \
  --protocol tcp \
  --destination 169.254.169.254 \
  --dport 80 \
  --jump DNAT \
  --in-interface docker0 \
  --to-destination 10.3.0.170:8181 \
  --out-interface flannel.1
```

`10.3.0.170` is the service IP. The service:

```
$ kubectl describe service/kube2iam
Name:			kube2iam
Namespace:		ricky-pai
Labels:			app=kube2iam
Selector:		name=kube2iam
Type:			ClusterIP
IP:			10.3.0.170
Port:			<unset>	8181/TCP
Endpoints:		10.0.0.209:8181,10.0.0.50:8181,10.0.0.52:8181
Session Affinity:	None
No events.
```

#### when crossing node boundaries, pod IP is lost
when I tried debugging by curling the service directly in command line in the container, pod IP is not always used.

Right now, have three nodes:

```
$ kubectl get nodes
NAME                         STATUS                     AGE
ip-10-0-0-209.ec2.internal    Ready                      1d
ip-10-0-0-50.ec2.internal    Ready,SchedulingDisabled   1d
ip-10-0-0-52.ec2.internal    Ready                      1d
```

each have one kube2iam running on hostNetwork through DaemonSet:
```
$ kubectl get pods -o=wide
NAME             READY     STATUS    RESTARTS   AGE       IP           NODE
kube2iam-c8ad7   1/1       Running   0          18h       10.0.0.52    ip-10-0-0-52.ec2.internal
kube2iam-vot7h   1/1       Running   0          18h       10.0.0.209   ip-10-0-0-209.ec2.internal
kube2iam-znuz2   1/1       Running   0          18h       10.0.0.50    ip-10-0-0-50.ec2.internal
```

when I do `curl 10.3.0.170:8181` inside a Pod,

the kube2iam instance on the same node will get the correct Pod IP:
```
$ kubectl logs -f kube2iam-c8ad7
time="2016-09-14T20:05:09Z" level=info msg="Requesting /" 
time="2016-09-14T20:05:09Z" level=debug msg="RemoteAddr 10.2.1.2" 
time="2016-09-14T20:05:09Z" level=debug msg="Proxied /" 
```

while the kube2iam instance on another node will get the node IP:
```
$ kubectl logs -f kube2iam-vot7h
time="2016-09-14T20:05:11Z" level=info msg="Requesting /" 
time="2016-09-14T20:05:11Z" level=debug msg="RemoteAddr 10.0.0.52" 
time="2016-09-14T20:05:11Z" level=debug msg="Proxied /" 
```

I also ran into the problem that the service never seems to reach kube2iam instances on master nodes when where curling the service IP within a pod. This might be problems with my own cluster setup though.

The second problem seems to prevent any possibility of running kube2iam on different instances as the requesting Pod. Suggestions and insights are greatly welcomed. I can also share more of my setup if needed.

**Comment by rickypai** [comment]
also, my setup uses flannel on vxlan. forgot to include that detail.

**Comment by jtblin** [comment]
Yeah I forgot about that obviously it uses the remote ip to determine from which pod it is originating and getting the role. One solution may be to run a proxy on the nodes and forward to the master setting `X-Forwarded-For` header and modify kube2iam to retrieve the ip from this header if present.

**Comment by rickypai** [comment]
that makes sense. so kube2iam would have two components:

- `kube2iam` running as static pods on master nodes
- `kube2iam-proxy` running as daemonset

then you'd have iptables rule to redirect traffic to metadata proxy to the `kube2iam-proxy`, which sends the pod ip to `kube2iam`, which sends back IAM credentials.

questions:
- I wonder if setting a proxy header is better vs. just have `kube2iam-proxy` talk to `kube2iam` via some http api. the former is simpler to implement, and the latter might open up more options.
- while this enables the minions to not require `sts:AssumeRole`, this does mean the requests to `kube2iam` can be spoofed.

**Comment by rickypai** [comment]
might be relevant: https://github.com/dump247/ec2metaproxy/issues/6

**Comment by jtblin** [comment]
> while this enables the minions to not require sts:AssumeRole, this does mean the requests to kube2iam can be spoofed.

Good point, we could pass a JWT via the Bearer authorization header as well which means dealing with public and private keys and make the installation fairly more complex.

**Comment by fiunchinho** [comment]
Any updates on this?

**Comment by erickt** [comment]
One feature I'd like the hypothetical `kube2iam-server` and `kube2iam-proxy` to have is to reduce the blast radius of a root level exploit against a worker. Presumably, if someone gets root, they can access the memory of any of the containers on a minion, and therefore get access to the AWS services that any of the containers can access. Unfortunately I'm not sure if there's anything we can do to protect against this.

However, We could pretty easily protect against an attacker using the server and proxy from being able to assume the role of a container outside of this minion. `kube2iam-server` should verify that the incoming requests for a pod IP are actually running on the minion that's running the proxy. For added security, we could also throw in a unique secret per proxy to make it even more difficult to exploit the system.

**Comment by jtblin** [comment]
Closing this issue as it's 8 years old.
